module/state_synchronization/execution_data_service.go

```diff
@@ -2,430 +2,44 @@ package state_synchronization
-	// Eventually, we will need to implement validation logic on Verification nodes to slash this behavior.
-
-	return v, bcr.TotalBytesRead(), nil
-}
-
-// Get gets the ExecutionData for the given root CID from the blobservice.
-// The returned error will be:
-// - MalformedDataError if some level of the blob tree cannot be properly deserialized
-// - BlobSizeLimitExceededError if any blob in the blob tree exceeds the maximum blob size
-// - BlobNotFoundError if some CID in the blob tree could not be found from the blobservice
-func (s *executionDataServiceImpl) Get(ctx context.Context, rootID flow.Identifier) (*ExecutionData, error) {
-	rootCid := flow.IdToCid(rootID)
-
-	logger := s.logger.With().Str("cid", rootCid.String()).Logger()
-	logger.Debug().Msg("getting execution data")
-
-	s.metrics.ExecutionDataGetStarted()
-
-	start := time.Now()
-	cids := []cid.Cid{rootCid}
-
-	var blobTreeSize uint64
-
-	for i := uint(0); i <= defaultMaxBlobTreeDepth; i++ {
-		v, totalBytes, err := s.getBlobs(ctx, cids, logger)
-
-		if err != nil {
-			s.metrics.ExecutionDataGetFinished(time.Since(start), false, blobTreeSize)
-
-			return nil, fmt.Errorf("failed to get level %d of blob tree: %w", i, err)
-		}
-
-		blobTreeSize += totalBytes
-
-		switch v := v.(type) {
-		case *ExecutionData:
-			s.metrics.ExecutionDataGetFinished(time.Since(start), true, blobTreeSize)
-
-			return v, nil
-		case *[]cid.Cid:
-			cids = *v
-		}
-	}
-
-	return nil, ErrBlobTreeDepthExceeded
-}
-
-// MalformedDataError is returned when malformed data is found at some level of the requested
-// blob tree. It likely indicates that the tree was generated incorrectly, and hence the request
-// should not be retried.
-type MalformedDataError struct {
-	err error
-}
-
-func (e *MalformedDataError) Error() string {
-	return fmt.Sprintf("malformed data: %v", e.err)
-}
-
-func (e *MalformedDataError) Unwrap() error { return e.err }
-
-// BlobSizeLimitExceededError is returned when a blob exceeds the maximum size allowed.
-type BlobSizeLimitExceededError struct {
-	cid cid.Cid
-}
-
-func (e *BlobSizeLimitExceededError) Error() string {
-	return fmt.Sprintf("blob %v exceeds maximum blob size", e.cid.String())
-}
-
-// BlobNotFoundError is returned when the blobservice failed to find a blob.
-type BlobNotFoundError struct {
-	cid cid.Cid
-}
-
-func (e *BlobNotFoundError) Error() string {
-	return fmt.Sprintf("blob %v not found", e.cid.String())
+const DefaultMaxBlobSize = 1 << 20 // 1MiB
+
+// ExecutionDataAdder handles adding execution data to a blobservice
+type ExecutionDataAdder interface {
+	// AddChunkExecutionData constructs a blob tree for the given ChunkExecutionData
+	// and adds it to the blobservice, and then returns the root CID.
+	AddChunkExecutionData(
+		ctx context.Context,
+		blockID flow.Identifier,
+		chunkIndex int,
+		ced *ChunkExecutionData,
+	) (cid.Cid, error)
```

Comment:
I'm not sure whether this method is necessary to be public. This might cause mistake from the caller who added 1 chunk, but didn't add the rest chunks. 

Instead we should just provide one method to add all chunks. So that there are only two results after the call, either nothing was added, or all chunks have been added.